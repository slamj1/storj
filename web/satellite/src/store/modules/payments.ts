// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import {
    BillingHistoryItem,
    BillingHistoryItemStatus,
    BillingHistoryItemType,
    CreditCard,
    DateRange,
    PaymentsApi,
    ProjectUsageAndCharges,
    TokenDeposit,
} from '@/types/payments';

export const PAYMENTS_MUTATIONS = {
    SET_BALANCE: 'SET_BALANCE',
    SET_CREDIT_CARDS: 'SET_CREDIT_CARDS',
    SET_DATE: 'SET_DATE',
    CLEAR: 'CLEAR_PAYMENT_INFO',
    UPDATE_CARDS_SELECTION: 'UPDATE_CARDS_SELECTION',
    UPDATE_CARDS_DEFAULT: 'UPDATE_CARDS_DEFAULT',
    SET_BILLING_HISTORY: 'SET_BILLING_HISTORY',
    SET_PROJECT_USAGE_AND_CHARGES: 'SET_PROJECT_USAGE_AND_CHARGES',
    SET_CURRENT_ROLLUP_PRICE: 'SET_CURRENT_ROLLUP_PRICE',
    SET_PREVIOUS_ROLLUP_PRICE: 'SET_PREVIOUS_ROLLUP_PRICE',
    SET_PRICE_SUMMARY: 'SET_PRICE_SUMMARY',
};

export const PAYMENTS_ACTIONS = {
    GET_BALANCE: 'getBalance',
    SETUP_ACCOUNT: 'setupAccount',
    GET_CREDIT_CARDS: 'getCreditCards',
    ADD_CREDIT_CARD: 'addCreditCard',
    CLEAR_PAYMENT_INFO: 'clearPaymentInfo',
    TOGGLE_CARD_SELECTION: 'toggleCardSelection',
    CLEAR_CARDS_SELECTION: 'clearCardsSelection',
    MAKE_CARD_DEFAULT: 'makeCardDefault',
    REMOVE_CARD: 'removeCard',
    GET_BILLING_HISTORY: 'getBillingHistory',
    MAKE_TOKEN_DEPOSIT: 'makeTokenDeposit',
    GET_PROJECT_USAGE_AND_CHARGES: 'getProjectUsageAndCharges',
    GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP: 'getProjectUsageAndChargesCurrentRollup',
    GET_PROJECT_USAGE_AND_CHARGES_PREVIOUS_ROLLUP: 'getProjectUsageAndChargesPreviousRollup',
};

const {
    SET_BALANCE,
    SET_CREDIT_CARDS,
    SET_DATE,
    CLEAR,
    UPDATE_CARDS_SELECTION,
    UPDATE_CARDS_DEFAULT,
    SET_BILLING_HISTORY,
    SET_PROJECT_USAGE_AND_CHARGES,
    SET_CURRENT_ROLLUP_PRICE,
    SET_PREVIOUS_ROLLUP_PRICE,
    SET_PRICE_SUMMARY,
} = PAYMENTS_MUTATIONS;

const {
    GET_BALANCE,
    SETUP_ACCOUNT,
    GET_CREDIT_CARDS,
    ADD_CREDIT_CARD,
    TOGGLE_CARD_SELECTION,
    CLEAR_CARDS_SELECTION,
    CLEAR_PAYMENT_INFO,
    MAKE_CARD_DEFAULT,
    REMOVE_CARD,
    GET_BILLING_HISTORY,
    MAKE_TOKEN_DEPOSIT,
    GET_PROJECT_USAGE_AND_CHARGES,
    GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP,
    GET_PROJECT_USAGE_AND_CHARGES_PREVIOUS_ROLLUP,
} = PAYMENTS_ACTIONS;

export class PaymentsState {
    /**
     * balance stores in cents
     */
    public balance: number = 0;
    public creditCards: CreditCard[] = [];
    public billingHistory: BillingHistoryItem[] = [];
    public usageAndCharges: ProjectUsageAndCharges[] = [];
    public priceSummary: number = 0;
    public currentRollupPrice: number = 0;
    public previousRollupPrice: number = 0;
    public startDate: Date = new Date();
    public endDate: Date = new Date();
}

/**
 * creates payments module with all dependencies
 *
 * @param api - payments api
 */
export function makePaymentsModule(api: PaymentsApi): StoreModule<PaymentsState> {
    return {
        state: new PaymentsState(),
        mutations: {
            [SET_BALANCE](state: PaymentsState, balance: number): void {
                state.balance = balance;
            },
            [SET_CREDIT_CARDS](state: PaymentsState, creditCards: CreditCard[]): void {
                state.creditCards = creditCards;
            },
            [SET_DATE](state: PaymentsState, dateRange: DateRange) {
                state.startDate = dateRange.startDate;
                state.endDate = dateRange.endDate;
            },
            [UPDATE_CARDS_SELECTION](state: PaymentsState, id: string | null): void {
                state.creditCards = state.creditCards.map(card => {
                    if (card.id === id) {
                        card.isSelected = !card.isSelected;

                        return card;
                    }

                    card.isSelected = false;

                    return card;
                });
            },
            [UPDATE_CARDS_DEFAULT](state: PaymentsState, id: string): void {
                state.creditCards = state.creditCards.map(card => {
                    if (card.id === id) {
                        card.isDefault = !card.isDefault;

                        return card;
                    }

                    card.isDefault = false;

                    return card;
                });
            },
            [SET_BILLING_HISTORY](state: PaymentsState, billingHistory: BillingHistoryItem[]): void {
                state.billingHistory = billingHistory;
            },
            [SET_PROJECT_USAGE_AND_CHARGES](state: PaymentsState, usageAndCharges: ProjectUsageAndCharges[]): void {
                state.usageAndCharges = usageAndCharges;
            },
            [SET_CURRENT_ROLLUP_PRICE](state: PaymentsState): void {
                state.currentRollupPrice = state.priceSummary;
            },
            [SET_PREVIOUS_ROLLUP_PRICE](state: PaymentsState): void {
                state.previousRollupPrice = state.priceSummary;
            },
            [SET_PRICE_SUMMARY](state: PaymentsState, charges: ProjectUsageAndCharges[]): void {
                if (charges.length === 0) {
                    state.priceSummary = 0;

                    return;
                }

                const usageItemSummaries = charges.map(item => item.summary());

                state.priceSummary = usageItemSummaries.reduce((accumulator, current) => accumulator + current);
            },
            [CLEAR](state: PaymentsState) {
                state.balance = 0;
                state.billingHistory = [];
                state.usageAndCharges = [];
                state.priceSummary = 0;
                state.previousRollupPrice = 0;
                state.currentRollupPrice = 0;
                state.creditCards = [];
                state.startDate = new Date();
                state.endDate = new Date();
            },
        },
        actions: {
            [GET_BALANCE]: async function({commit}: any): Promise<number> {
                const balance = await api.getBalance();

                commit(SET_BALANCE, balance);

                return balance;
            },
            [SETUP_ACCOUNT]: async function(): Promise<void> {
                await api.setupAccount();
            },
            [GET_CREDIT_CARDS]: async function({commit}: any): Promise<CreditCard[]> {
                const creditCards = await api.listCreditCards();

                commit(SET_CREDIT_CARDS, creditCards);

                return creditCards;
            },
            [ADD_CREDIT_CARD]: async function({commit}: any, token: string): Promise<void> {
                await api.addCreditCard(token);
            },
            [TOGGLE_CARD_SELECTION]: function({commit}: any, id: string): void {
                commit(UPDATE_CARDS_SELECTION, id);
            },
            [CLEAR_CARDS_SELECTION]: function({commit}: any): void {
                commit(UPDATE_CARDS_SELECTION, null);
            },
            [MAKE_CARD_DEFAULT]: async function({commit}: any, id: string): Promise<void> {
                await api.makeCreditCardDefault(id);

                commit(UPDATE_CARDS_DEFAULT, id);
            },
            [REMOVE_CARD]: async function({commit, state}: any, cardId: string): Promise<void> {
                await api.removeCreditCard(cardId);

                commit(SET_CREDIT_CARDS, state.creditCards.filter(card => card.id !== cardId));
            },
            [CLEAR_PAYMENT_INFO]: function({commit}: any): void {
                commit(CLEAR);
            },
            [GET_BILLING_HISTORY]: async function({commit}: any): Promise<void> {
                const billingHistory: BillingHistoryItem[] = await api.billingHistory();

                commit(SET_BILLING_HISTORY, billingHistory);
            },
            [MAKE_TOKEN_DEPOSIT]: async function({commit}: any, amount: number): Promise<TokenDeposit> {
                return await api.makeTokenDeposit(amount);
            },
            [GET_PROJECT_USAGE_AND_CHARGES]: async function({commit}: any, dateRange: DateRange): Promise<void> {
                const now = new Date();
                let beforeUTC = new Date(Date.UTC(dateRange.endDate.getUTCFullYear(), dateRange.endDate.getUTCMonth(), dateRange.endDate.getUTCDate(), 23, 59));

                if (now.getUTCFullYear() === dateRange.endDate.getUTCFullYear() &&
                    now.getUTCMonth() === dateRange.endDate.getUTCMonth() &&
                    now.getUTCDate() <= dateRange.endDate.getUTCDate()) {
                    beforeUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate(), now.getUTCHours(), now.getUTCMinutes()));
                }

                const sinceUTC = new Date(Date.UTC(dateRange.startDate.getUTCFullYear(), dateRange.startDate.getUTCMonth(), dateRange.startDate.getUTCDate(), 0, 0));
                const usageAndCharges: ProjectUsageAndCharges[] = await api.projectsUsageAndCharges(sinceUTC, beforeUTC);

                commit(SET_DATE, dateRange);
                commit(SET_PROJECT_USAGE_AND_CHARGES, usageAndCharges);
                commit(SET_PRICE_SUMMARY, usageAndCharges);
            },
            [GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP]: async function({commit}: any): Promise<void> {
                const now = new Date();
                const endUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate(), now.getUTCHours(), now.getUTCMinutes()));
                const startUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), 1, 0, 0));

                const usageAndCharges: ProjectUsageAndCharges[] = await api.projectsUsageAndCharges(startUTC, endUTC);

                commit(SET_DATE, new DateRange(startUTC, endUTC));
                commit(SET_PROJECT_USAGE_AND_CHARGES, usageAndCharges);
                commit(SET_PRICE_SUMMARY, usageAndCharges);
                commit(SET_CURRENT_ROLLUP_PRICE);
            },
            [GET_PROJECT_USAGE_AND_CHARGES_PREVIOUS_ROLLUP]: async function({commit}: any): Promise<void> {
                const now = new Date();
                const startUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth() - 1, 1, 0, 0));
                const endUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), 0, 23, 59, 59));

                const usageAndCharges: ProjectUsageAndCharges[] = await api.projectsUsageAndCharges(startUTC, endUTC);

                commit(SET_DATE, new DateRange(startUTC, endUTC));
                commit(SET_PROJECT_USAGE_AND_CHARGES, usageAndCharges);
                commit(SET_PRICE_SUMMARY, usageAndCharges);
                commit(SET_PREVIOUS_ROLLUP_PRICE);
            },
        },
        getters: {
            canUserCreateFirstProject: (state: PaymentsState): boolean => {
                return (state.billingHistory.some((billingItem: BillingHistoryItem) => {
                    return billingItem.amount >= 50 && billingItem.type === BillingHistoryItemType.Transaction
                        && billingItem.status === BillingHistoryItemStatus.Completed;
                }) && state.balance > 0) || state.creditCards.length > 0;
            },
            isTransactionProcessing: (state: PaymentsState): boolean => {
                return state.billingHistory.some((billingItem: BillingHistoryItem) => {
                    return billingItem.amount >= 50 && billingItem.type === BillingHistoryItemType.Transaction
                        && (billingItem.status === BillingHistoryItemStatus.Pending
                        || billingItem.status === BillingHistoryItemStatus.Paid
                        || billingItem.status === BillingHistoryItemStatus.Completed);
                }) && state.balance === 0;
            },
            isTransactionCompleted: (state: PaymentsState): boolean => {
                return (state.billingHistory.some((billingItem: BillingHistoryItem) => {
                    return billingItem.amount >= 50 && billingItem.type === BillingHistoryItemType.Transaction
                        && billingItem.status === BillingHistoryItemStatus.Completed;
                }) && state.balance > 0);
            },
            isInvoiceForPreviousRollup: (state: PaymentsState): boolean => {
                const now = new Date();

                return state.billingHistory.some((billingItem: BillingHistoryItem) => {
                    if (now.getUTCMonth() === 0) {
                        return billingItem.type === BillingHistoryItemType.Invoice
                            && billingItem.start.getUTCFullYear() === now.getUTCFullYear() - 1
                            && billingItem.start.getUTCMonth() === 11;
                    }

                    return billingItem.type === BillingHistoryItemType.Invoice
                        && billingItem.start.getUTCFullYear() === now.getUTCFullYear()
                        && billingItem.start.getUTCMonth() === now.getUTCMonth() - 1;
                });
            },
        },
    };
}
