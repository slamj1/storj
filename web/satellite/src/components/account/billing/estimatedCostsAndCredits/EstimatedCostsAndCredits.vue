// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="current-month-area">
        <div class="current-month-area__title-area">
            <h1 class="current-month-area__title-area__title">Estimated Costs for This Billing Period</h1>
            <span class="current-month-area__title-area__costs">{{ priceSummary | centsToDollars }}</span>
        </div>
        <div class="current-month-area__content">
            <h2 class="current-month-area__content__title">DETAILS</h2>
            <div class="current-month-area__content__usage-charges" @click="toggleUsageChargesPopup">
                <div class="current-month-area__content__usage-charges__head">
                    <div class="current-month-area__content__usage-charges__head__name-area">
                        <div class="current-month-area__content__usage-charges__head__name-area__image-container" v-if="projectUsageAndCharges.length > 0">
                            <ArrowRightIcon v-if="!areProjectUsageAndChargesShown"/>
                            <ArrowDownIcon v-else/>
                        </div>
                        <span class="current-month-area__content__usage-charges__head__name-area__title">Usage Charges</span>
                    </div>
                    <span>Estimated total <span class="summary">{{ priceSummary | centsToDollars }}</span></span>
                </div>
                <div class="current-month-area__content__usage-charges__content" v-if="areProjectUsageAndChargesShown" @click.stop>
                    <UsageAndChargesItem
                        v-for="usageAndCharges in projectUsageAndCharges"
                        :item="usageAndCharges"
                        :key="usageAndCharges.projectId"
                        class="item"
                    />
                </div>
            </div>
            <div class="current-month-area__content__credits-area">
                <div class="current-month-area__content__credits-area__title-area">
                    <span class="current-month-area__content__credits-area__title-area__title">Earned Credits</span>
                </div>
                <span
                    :style="{ color: balanceColor }"
                    class="current-month-area__content__credits-area__balance"
                >
                    {{ balance | centsToDollars }}
                </span>
            </div>
<!--            <div class="current-month-area__content__credits-area">-->
<!--                <div class="current-month-area__content__credits-area__title-area">-->
<!--                    <span class="current-month-area__content__credits-area__title-area__title">Available Credits</span>-->
<!--                </div>-->
<!--                <span class="current-month-area__content__credits-area__balance">-->
<!--                    {{ availableBalance | centsToDollars }}-->
<!--                </span>-->
<!--            </div>-->
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import UsageAndChargesItem from '@/components/account/billing/estimatedCostsAndCredits/UsageAndChargesItem.vue';
import VButton from '@/components/common/VButton.vue';

import ArrowRightIcon from '@/../static/images/common/BlueArrowRight.svg';
import ArrowDownIcon from '@/../static/images/common/BlueExpand.svg';

import { PAYMENTS_ACTIONS } from '@/store/modules/payments';
import { ProjectUsageAndCharges } from '@/types/payments';

@Component({
    components: {
        VButton,
        UsageAndChargesItem,
        ArrowRightIcon,
        ArrowDownIcon,
    },
})
export default class EstimatedCostsAndCredits extends Vue {
    /**
     * Lifecycle hook after initial render.
     * Fetches current project usage rollup.
     */
    public async mounted(): Promise<void> {
        try {
            await this.$store.dispatch(PAYMENTS_ACTIONS.GET_BALANCE);
            await this.$store.dispatch(PAYMENTS_ACTIONS.GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP);
        } catch (error) {
            await this.$notify.error(error.message);
        }
    }

    /**
     * areProjectUsageAndChargesShown indicates if area with all projects is expanded.
     */
    public areProjectUsageAndChargesShown: boolean = false;

    /**
     * projectUsageAndCharges is an array of all stored ProjectUsageAndCharges.
     */
    public get projectUsageAndCharges(): ProjectUsageAndCharges[] {
        return this.$store.state.paymentsModule.usageAndCharges;
    }

    /**
     * priceSummary returns price summary of usages for all the projects.
     */
    public get priceSummary(): number {
        return this.$store.state.paymentsModule.priceSummary;
    }

    /**
     * Returns balance from store in cents.
     */
    public get balance(): number {
        return this.$store.state.paymentsModule.balance;
    }

    // TODO: use when coupon expiration bug is fixed
    // /**
    //  * Returns available balance in cents.
    //  */
    // public get availableBalance(): number {
    //     const total = this.previousRollupPrice + this.currentRollupPrice;
    //
    //     switch (true) {
    //         case this.balance <= total:
    //             return 0;
    //         case this.$store.getters.isInvoiceForPreviousRollup:
    //             return this.balance - this.currentRollupPrice;
    //         default:
    //             return this.balance - total;
    //     }
    // }

    /**
     * Returns balance color red if balance below zero and clack if not.
     */
    public get balanceColor(): string {
        return this.$store.state.paymentsModule.balance < 0 ? '#FF0000' : '#000';
    }

    /**
     * toggleUsageChargesPopup is used to open/close area with list of project charges.
     */
    public toggleUsageChargesPopup(): void {
        if (this.projectUsageAndCharges.length === 0) {
            return;
        }

        this.areProjectUsageAndChargesShown = !this.areProjectUsageAndChargesShown;
    }

    // TODO: use when coupon expiration bug is fixed
    // /**
    //  * previousRollupPrice is a price of previous rollup.
    //  */
    // private get previousRollupPrice(): number {
    //     return this.$store.state.paymentsModule.previousRollupPrice;
    // }
    //
    // /**
    //  * currentRollupPrice is a price of current rollup.
    //  */
    // private get currentRollupPrice(): number {
    //     return this.$store.state.paymentsModule.currentRollupPrice;
    // }
}
</script>

<style scoped lang="scss">
    h1,
    h2,
    p,
    span {
        margin: 0;
        color: #354049;
    }

    .current-month-area {
        margin-bottom: 32px;
        padding: 40px;
        background-color: #fff;
        border-radius: 8px;
        font-family: 'font_regular', sans-serif;

        &__title-area {
            display: flex;
            justify-content: space-between;
            align-items: center;
            padding-bottom: 40px;
            border-bottom: 1px solid rgba(169, 181, 193, 0.3);

            &__title,
            &__costs {
                font-size: 28px;
                line-height: 42px;
                font-family: 'font_bold', sans-serif;
                color: #354049;
            }
        }

        &__content {
            margin-top: 20px;

            &__title {
                font-size: 16px;
                line-height: 23px;
                letter-spacing: 0.04em;
                text-transform: uppercase;
                color: #919191;
            }

            &__usage-charges {
                position: relative;
                margin: 18px 0 0 0;
                background-color: #f5f6fa;
                border-radius: 12px;
                cursor: pointer;

                &__head {
                    display: flex;
                    justify-content: space-between;
                    align-items: center;
                    padding: 20px;

                    &__name-area {
                        display: flex;
                        align-items: center;

                        &__image-container {
                            max-width: 14px;
                            max-height: 14px;
                            width: 14px;
                            height: 14px;
                            display: flex;
                            align-items: center;
                            justify-content: center;
                            margin-right: 12px;
                        }
                    }
                }

                &__content {
                    cursor: default;
                    max-height: 228px;
                    overflow-y: auto;
                    padding: 0 20px;
                }
            }

            &__credits-area {
                display: flex;
                align-items: center;
                justify-content: space-between;
                padding: 20px;
                width: calc(100% - 40px);
                background-color: #f5f6fa;
                border-radius: 12px;
                margin-top: 20px;

                &__title-area {
                    display: flex;
                    align-items: center;

                    &__title {
                        font-size: 16px;
                        line-height: 21px;
                        color: #354049;
                    }
                }
            }
        }
    }

    .item {
        border-top: 1px solid rgba(169, 181, 193, 0.3);
    }

    .summary {
        user-select: text;
    }
</style>
