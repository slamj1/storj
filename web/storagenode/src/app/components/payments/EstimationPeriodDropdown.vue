// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="period-container" @click.stop="openPeriodDropdown">
        <p class="period-container__label">{{ currentPeriod }}</p>
        <BlackArrowHide v-if="isCalendarShown" />
        <BlackArrowExpand v-else />
        <PayoutPeriodCalendar
            class="period-container__calendar"
            v-click-outside="closePeriodDropdown"
            v-if="isCalendarShown"
        />
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import PayoutPeriodCalendar from '@/app/components/payments/PayoutPeriodCalendar.vue';

import BlackArrowExpand from '@/../static/images/BlackArrowExpand.svg';
import BlackArrowHide from '@/../static/images/BlackArrowHide.svg';

import { APPSTATE_ACTIONS } from '@/app/store/modules/appState';
import { PayoutPeriod } from '@/app/types/payout';

/**
 * Holds all months names.
 */
const monthNames = [
    'January', 'February', 'March', 'April',
    'May', 'June', 'July',	'August',
    'September', 'October', 'November',	'December',
];

@Component({
    components: {
        PayoutPeriodCalendar,
        BlackArrowExpand,
        BlackArrowHide,
    },
})
export default class EstimationPeriodDropdown extends Vue {
    /**
     * Returns formatted selected payout period.
     */
    public get currentPeriod(): string {
        const start: PayoutPeriod = this.$store.state.payoutModule.periodRange.start;
        const end: PayoutPeriod = this.$store.state.payoutModule.periodRange.end;

        return start && start.period !== end.period ?
            `${monthNames[start.month]}, ${start.year} - ${monthNames[end.month]}, ${end.year}`
            : `${monthNames[end.month]}, ${end.year}`;
    }

    /**
     * Indicates if period selection calendar should appear.
     */
    public get isCalendarShown(): boolean {
        return this.$store.state.appStateModule.isPayoutCalendarShown;
    }

    /**
     * Opens payout period selection dropdown.
     */
    public openPeriodDropdown(): void {
        this.$store.dispatch(APPSTATE_ACTIONS.TOGGLE_PAYOUT_CALENDAR, true);
    }

    /**
     * Closes payout period selection dropdown.
     */
    public closePeriodDropdown(): void {
        this.$store.dispatch(APPSTATE_ACTIONS.TOGGLE_PAYOUT_CALENDAR, false);
    }
}
</script>

<style scoped lang="scss">
    .period-container {
        position: relative;
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: center;
        background-color: transparent;
        cursor: pointer;

        &__label {
            margin-right: 8px;
            font-family: 'font_regular', sans-serif;
            font-weight: 500;
            font-size: 16px;
            color: var(--month-label-color);
        }

        &__calendar {
            position: absolute;
            top: 30px;
            right: 0;
        }
    }

    .arrow {

        path {
            fill: var(--period-selection-arrow-color);
        }
    }
</style>
