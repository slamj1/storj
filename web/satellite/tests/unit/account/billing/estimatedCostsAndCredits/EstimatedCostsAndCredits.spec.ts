// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import EstimatedCostsAndCredits from '@/components/account/billing/estimatedCostsAndCredits/EstimatedCostsAndCredits.vue';

import { makePaymentsModule, PAYMENTS_MUTATIONS } from '@/store/modules/payments';
import { makeProjectsModule, PROJECTS_MUTATIONS } from '@/store/modules/projects';
import { makeUsersModule, USER_MUTATIONS } from '@/store/modules/users';
import { BillingHistoryItem, BillingHistoryItemType, ProjectUsageAndCharges } from '@/types/payments';
import { Project } from '@/types/projects';
import { User } from '@/types/users';
import { createLocalVue, mount } from '@vue/test-utils';

import { PaymentsMock } from '../../../mock/api/payments';
import { ProjectsApiMock } from '../../../mock/api/projects';
import { UsersApiMock } from '../../../mock/api/users';

const localVue = createLocalVue();
localVue.filter('centsToDollars', (cents: number): string => {
    return `USD $${(cents / 100).toFixed(2)}`;
});
localVue.use(Vuex);

const usersApi = new UsersApiMock();
const usersModule = makeUsersModule(usersApi);
const projectsApi = new ProjectsApiMock();
const projectsModule = makeProjectsModule(projectsApi);
const paymentsApi = new PaymentsMock();
const paymentsModule = makePaymentsModule(paymentsApi);
const store = new Vuex.Store({ modules: { usersModule, projectsModule, paymentsModule }});

const project = new Project('id', 'projectName', 'projectDescription', 'test', 'testOwnerId', true);
const user = new User('testOwnerId');
const date = new Date(1970, 1, 1);
const projectCharge = new ProjectUsageAndCharges(date, date, 100, 100, 100, 'id', 100, 100, 100);
const {
    SET_BALANCE,
    CLEAR,
    SET_BILLING_HISTORY,
    SET_PROJECT_USAGE_AND_CHARGES,
    SET_CURRENT_ROLLUP_PRICE,
    SET_PREVIOUS_ROLLUP_PRICE,
    SET_PRICE_SUMMARY,
} = PAYMENTS_MUTATIONS;

describe('EstimatedCostsAndCredits', () => {
    it('renders correctly with project and no project usage and charges', () => {
        store.commit(USER_MUTATIONS.SET_USER, user);
        store.commit(PROJECTS_MUTATIONS.ADD, project);

        const wrapper = mount(EstimatedCostsAndCredits, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with project and project usage and charges', () => {
        store.commit(SET_BALANCE, 5500);
        store.commit(SET_PROJECT_USAGE_AND_CHARGES, [projectCharge]);
        store.commit(SET_PRICE_SUMMARY, [projectCharge]);
        store.commit(SET_CURRENT_ROLLUP_PRICE);
        store.commit(SET_PREVIOUS_ROLLUP_PRICE);

        const wrapper = mount(EstimatedCostsAndCredits, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });

    // TODO: use when coupon expiration bug is fixed
    // it('renders correctly with project and project usage and charges with previous rollup invoice', () => {
    //     const now = new Date();
    //     let billingHistoryItem = new BillingHistoryItem('id', 'description', 300, 300, 'paid', 'test', new Date(now.getUTCFullYear(), now.getUTCMonth() - 1, 15), now, BillingHistoryItemType.Invoice);
    //
    //     if (now.getUTCMonth() === 0) {
    //         billingHistoryItem = new BillingHistoryItem('id', 'description', 300, 300, 'paid', 'test', new Date(now.getUTCFullYear() - 1, 11, 15), now, BillingHistoryItemType.Invoice);
    //     }
    //
    //     store.commit(CLEAR);
    //     store.commit(SET_BALANCE, 600);
    //     store.commit(SET_PROJECT_USAGE_AND_CHARGES, [projectCharge]);
    //     store.commit(SET_PRICE_SUMMARY, [projectCharge]);
    //     store.commit(SET_CURRENT_ROLLUP_PRICE);
    //     store.commit(SET_BILLING_HISTORY, [billingHistoryItem]);
    //
    //     const wrapper = mount(EstimatedCostsAndCredits, {
    //         store,
    //         localVue,
    //     });
    //
    //     expect(wrapper).toMatchSnapshot();
    // });
    //
    // it('renders correctly with project and project usage and charges with price bigger than balance amount', () => {
    //     store.commit(CLEAR);
    //     store.commit(SET_BALANCE, 500);
    //     store.commit(SET_PROJECT_USAGE_AND_CHARGES, [projectCharge]);
    //     store.commit(SET_PRICE_SUMMARY, [projectCharge]);
    //     store.commit(SET_CURRENT_ROLLUP_PRICE);
    //     store.commit(SET_PREVIOUS_ROLLUP_PRICE);
    //     store.commit(SET_BILLING_HISTORY, []);
    //
    //     const wrapper = mount(EstimatedCostsAndCredits, {
    //         store,
    //         localVue,
    //     });
    //
    //     expect(wrapper).toMatchSnapshot();
    // });
});
