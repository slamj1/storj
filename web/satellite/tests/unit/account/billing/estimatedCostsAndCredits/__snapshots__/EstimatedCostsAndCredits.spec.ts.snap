// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EstimatedCostsAndCredits renders correctly with project and no project usage and charges 1`] = `
<div class="current-month-area">
  <div class="current-month-area__title-area">
    <h1 class="current-month-area__title-area__title">Estimated Costs for This Billing Period</h1> <span class="current-month-area__title-area__costs">USD $0.00</span>
  </div>
  <div class="current-month-area__content">
    <h2 class="current-month-area__content__title">DETAILS</h2>
    <div class="current-month-area__content__usage-charges">
      <div class="current-month-area__content__usage-charges__head">
        <div class="current-month-area__content__usage-charges__head__name-area">
          <!----> <span class="current-month-area__content__usage-charges__head__name-area__title">Usage Charges</span></div> <span>Estimated total <span class="summary">USD $0.00</span></span>
      </div>
      <!---->
    </div>
    <div class="current-month-area__content__credits-area">
      <div class="current-month-area__content__credits-area__title-area"><span class="current-month-area__content__credits-area__title-area__title">Earned Credits</span></div> <span class="current-month-area__content__credits-area__balance" style="color: rgb(0, 0, 0);">
                    USD $0.00
                </span>
    </div>
  </div>
</div>
`;

exports[`EstimatedCostsAndCredits renders correctly with project and project usage and charges 1`] = `
<div class="current-month-area">
  <div class="current-month-area__title-area">
    <h1 class="current-month-area__title-area__title">Estimated Costs for This Billing Period</h1> <span class="current-month-area__title-area__costs">USD $3.00</span>
  </div>
  <div class="current-month-area__content">
    <h2 class="current-month-area__content__title">DETAILS</h2>
    <div class="current-month-area__content__usage-charges">
      <div class="current-month-area__content__usage-charges__head">
        <div class="current-month-area__content__usage-charges__head__name-area">
          <div class="current-month-area__content__usage-charges__head__name-area__image-container"><svg width="8" height="14" viewBox="0 0 8 14" fill="none" xmlns="http://www.w3.org/2000/svg">
              <path fill-rule="evenodd" clip-rule="evenodd" d="M0.363528 13.6272C-0.121176 13.1302 -0.121176 12.3243 0.363528 11.8273L5.07113 7L0.363528 2.17268C-0.121177 1.67565 -0.121177 0.869804 0.363528 0.372774C0.848232 -0.124258 1.63409 -0.124258 2.1188 0.372774L8.58166 7L2.1188 13.6272C1.63409 14.1243 0.848232 14.1243 0.363528 13.6272Z" fill="#2683FF"></path>
            </svg></div> <span class="current-month-area__content__usage-charges__head__name-area__title">Usage Charges</span>
        </div> <span>Estimated total <span class="summary">USD $3.00</span></span>
      </div>
      <!---->
    </div>
    <div class="current-month-area__content__credits-area">
      <div class="current-month-area__content__credits-area__title-area"><span class="current-month-area__content__credits-area__title-area__title">Earned Credits</span></div> <span class="current-month-area__content__credits-area__balance" style="color: rgb(0, 0, 0);">
                    USD $55.00
                </span>
    </div>
  </div>
</div>
`;
